Adjust docs to account for SHConfig changes (#443)
sentinel-hub/sentinelhub-py: docs/source/configure.rst

Configuration
=============

The package contains a configuration file ``config.json``. After the package is installed you can check the initial
configuration parameters using the command line::

> $ sentinelhub.config --show

Reviewer: This should probably stay in, right?

Author: the `--show` functionality?

Reviewer: Yes, unless it was removed, now I don't remember anymore.

> Part of the package configuration is represented by the :class:`~sentinelhub.config.SHConfig` class. It can be adjusted and passed to most of the functions and constructors of the ``sentinelhub`` package to either provide credentials or modify behavior such as the number download processes.

Reviewer: ```suggestion
Part of the package configuration is represented by the :class:`~sentinelhub.config.SHConfig` class. It can be adjusted and passed to most of the functions and constructors of the ``sentinelhub`` package to either provide credentials or modify behavior such as the number of download processes.
```

    >    instance_id='',
    >    sh_client_id='',
    >    sh_client_secret='',
    >    sh_base_url='https://services.sentinel-hub.com',
    >    sh_auth_base_url='https://services.sentinel-hub.com',
    >    ...
    > )

To avoid the need of constant reconfiguration in the code, we also support adjusting the values in a ``config.toml`` file.

Configuration File
******************

> Whenever a new ``SHConfig`` object is created, the default values of the fields are updated with the contents of the configuration file. The configuration file also supports multiple profiles, which can be used with ``SHConfig("myprofile")``. If no profile is specified, the default profile is used (``sentinelhub.config.DEFAULT_PROFILE``, at the time of writing this is ``"default-profile"``). This is also used whenever no explicit ``SHConfig`` was provided to a function/class.

Reviewer: ```suggestion
Whenever a new ``SHConfig`` object is created, the default values of the fields are updated with the contents of the configuration file. The configuration file also supports multiple profiles, which can be used with ``SHConfig("myprofile")``. If no profile is specified, the default profile is used (``sentinelhub.config.DEFAULT_PROFILE``, currently set to ``"default-profile"``). This is also used whenever no explicit ``SHConfig`` is provided to a function/class, unless the preferred profile is set via the `SH_PROFILE` environment variable (more about that in a later section).
```

> The configuration file can be found at ``~/.config/sentinelhub/config.toml``. On Windows that usually translates to ``C:/Users/<USERNAME>/.config/sentinelhub/config.toml``. You can get the precise location of the file by calling ``SHConfig.get_config_location()``.

Reviewer: ```suggestion
The configuration file can be found at ``~/.config/sentinelhub/config.toml``. On Windows this usually translates to ``C:/Users/<USERNAME>/.config/sentinelhub/config.toml``. You can get the precise location of the file by calling ``SHConfig.get_config_location()``.
```

> The file follows the standard TOML structure. The sections are denoted by profile name and the following lines specify ``key=value`` pairs for any fields that should be updated, for example:

Reviewer: ```suggestion
The configuration file follows the standard TOML structure. Sections are denoted by the profile name in square brackets, while the following lines specify ``key=value`` pairs for any fields that should be updated, for example:
```

.. code-block:: toml

    [default-profile]
    instance_id = "my-instance-id"
    max_download_attempts = 3

    [custom-profile]
    instance_id = "something-else"

The file can also be updated programmatically by using the ``save`` method.

.. code-block:: python

    from sentinelhub import SHConfig

    config = SHConfig()
    config.instance_id = "my-instance-id"
    config.save("my-profile")

Another option is to update the configuration file via CLI. However this approach can only modify existing profiles, any new profiles need to be added manually or through the Python interface.::

$ sentinelhub.config --profile my-profile --instance_id my-instance-id

Environment Variables
*********************

We generally suggest using the configuration file, but we offer limited support for environmental variables to simplify situations such as building docker images.

The ``SHConfig`` class reads the following environmental variables during initialization:

> - ``SH_PROFILE`` that dictates which profile is loaded if not explicitly provided.

Reviewer: ```suggestion
- ``SH_PROFILE`` that dictates which profile should be used when not explicitly provided.
```

- ``SH_CLIENT_ID`` and ``SH_CLIENT_SECRET`` for setting the SentinelHub credentials.

Precedence
**********

The general precedence order is ``explicit parameters > environment > configuration file > defaults``.

This means that ``SHConfig(profile="my-profile", sh_client_id="my-id")`` will be taken into account over ``SH_PROFILE`` and ``SH_CLIENT_ID`` environment variables, which would take precedence over what is specified in the ``configuration.toml``.

> Sentinel Hub Configuration

Reviewer: I feel like this should be moved up, because it describes a more general overview of the configuration and offers some explanations. Perhaps at the top, even?

The AWS configuration could stay below, since it's a specific use case

Author: but this is only relevant to three of the parameters no?

Reviewer: you're correct regarding the parameters POV, but the rest of the text reads like an intro which gives a broader meaning to the whole point of configuration. 

I'm not married to the idea, so if you don't like it, feel free to ignore.

**************************

In order to use Sentinel Hub services you will need a Sentinel Hub account. If you do not have one yet, you can
create a free trial account at `Sentinel Hub`_. If you are a researcher you can even apply for a free non-commercial
account at `ESA OSEO page`_. The following configurations are then linked to your account.

Parameter ``instance_id`` is used when using OGC endpoints of the `Sentinel Hub services`_. It is the identifier of a
configuration users can set up in the `Sentinel Hub Dashboard`_ under "Configuration Utility".

The ``sh_client_id`` and ``sh_client_secret`` parameters can also be created in the `Sentinel Hub Dashboard`_ under
"User settings". The two parameters are needed when accessing protected endpoints of the service (Process, Catalog,
Batch, BYOC, and other APIs). There is "OAuth clients" frame where we can create a new OAuth client.

.. admonition:: Additional information on creating OAuth client

    For detailed instructions on how to obtain credentials, you can see the `Sentinel Hub webinar`_.

> Amazon S3 Configuration
> ***********************
> 
> The package enables downloading Sentinel-2 L1C and L2A data from `Amazon S3`_ storage buckets. The data is contained in
> Requester Pays buckets therefore `AWS credentials`_ are required to use these capabilities. The credentials
> can be set in the package configuration file with parameters ``aws_access_key_id`` and ``aws_secret_access_key``.

Reviewer: ```suggestion
Amazon S3 Configuration
***********************

The package enables downloading Sentinel-2 L1C and L2A data from `Amazon S3`_ storage buckets. The data is contained in
Requester Pays buckets, therefore `AWS credentials`_ are required to use these capabilities. The credentials
can be set in the package configuration file with parameters ``aws_access_key_id`` and ``aws_secret_access_key``.
```
